### DISCUSSION TITLE
Training Data Sample for NER

### DISCUSSION THREAD

**Discussion by aniyyanz08** [explosion/spaCy#9495] [Help: Model Advice] [answered]
I am trying to build a custom NER model. The data is scraped from internet and cleaned. I have several thousands of these type of data for training purpose. I have annotated this data using annotation tools and one sample of data in spacy training format is given.

```
[('A Very SoNA Christmas\nView SoNA’s Covid Safety Policies\nSkip to Content\nAbout\nHistory Mission\nStaff Board\nMusic Director\nMusicians\nSoNA Singers\nAuditions\nHire Ensembles\nContact\n2021-22 Season\nSubscriber Series\nTicketed Performances\nSoNA Beyond Series\nVirtual Performances\nVirtual Performances\nSolos from Home\nSpecial Events\nFireworks at the Farm\nReimagined Celebration\nDonate\nGallery\nEducation\nBlog\nOpen Menu\nClose Menu\nAbout\nHistory Mission\nStaff Board\nMusic Director\nMusicians\nSoNA Singers\nAuditions\nHire Ensembles\nContact\n2021-22 Season\nSubscriber Series\nTicketed Performances\nSoNA Beyond Series\nVirtual Performances\nVirtual Performances\nSolos from Home\nSpecial Events\nFireworks at the Farm\nReimagined Celebration\nDonate\nGallery\nEducation\nBlog\nOpen Menu\nClose Menu\nFolder:\nAbout\nFolder:\n2021-22 Season\nSoNA Beyond Series\nFolder:\nVirtual Performances\nFolder:\nSpecial Events\nDonate\nGallery\nEducation\nBlog\nBack\nHistory Mission\nStaff Board\nMusic Director\nMusicians\nSoNA Singers\nAuditions\nHire Ensembles\nContact\nBack\nSubscriber Series\nTicketed Performances\nBack\nVirtual Performances\nSolos from Home\nBack\nFireworks at the Farm\nReimagined Celebration\nA Very SoNA Christmas\nJul 10, 2021\nWritten By SoNA\nSaturday, December 11, 2021 2PM 7:30PM Walton Arts Center, Fayetteville\nA mix of sacred and secular holiday favorites with local guest soloists, The SoNA Singers, and area high school and collegiate choruses. Saturday, December 11, 2021 2PM Matinee Performance Saturday, December 11, 2021 7:30PM Evening Performance\nBuy Tickets\nBuy Tickets\nSingle Tickets: 35, 45, 57 Under 18 FREE with purchase of adult ticket limited quantities Interested in a full season subscription Learn more here . Concert sponsored by Bogle Family Foundation\nWe are committed to ensuring that audiences can experience music safely in person at our performances. Until further notice, patrons, staff, and volunteers are required to wear masks. Learn more about our safety policy here .\nSoNA\nPrevious\nPrevious\nMozart and Beethoven\nNext\nNext\nSoNA Walton Arts Center present The Snowman: A Family Concert\nReceive the latest updates\nEmail Address\nSign Up\nThank you for joining our email list You should receive a verification email shortly to confirm.\nOffice: 479.521.4166 Tickets: 479.443.5600 infosonamusic.org\nCopyright 2021, SoNA. All rights reserved.\nSupport SoNA',
  {'entities': [(1958, 1962, 'organization'),
    (1230, 1236, 'performance_starttime'),
    (1343, 1359, 'organization'),
    (1208, 1225, 'performance_date'),
    (1237, 1255, 'auditorium'),
    (0, 21, 'production_name'),
    (1226, 1229, 'performance_starttime')]})]
```
I would like to know if i take this much bigger unstructured data, will NER is going to work. Also any insight for improvement from the above annotated data is requested.

**Comment by polm** [comment, marked as answer]
You can use documents of that length, but in general it's easier to work with documents if you cut them down to paragraph length. In your case there don't seem to be real paragraphs but it looks like you could split the data into lines without losing information. 

There are other issues with your data. A lot of lines are irrelevant and could be pre-filtered, which would make the rest of your task much easier. You can filter lines by removing set phrases or overly short lines. 

The show title is the first line, but there's not any useful context for the model to learn, or any useful keywords really. So if most of your data looks like that it won't help.

You're labelling the show start time and date, but those are the only dates in the text except for years (season and copyright), which can be filtered out easily. So you should be able to just use the date label in the pretrained model.

**Reply by aniyyanz08** [reply]
This is the data from one url. There is more than 5000 data from different urls which follow different pattern and length. I thought of considering data from one url as a single data and from other urls as different data. I need to extract similar entities ( not just these) like artist name, role etc etc and most of them dont follow a line format as we can see in the website.  Also how can i remove the lines which are irrelevant and filter those lines. If i do that then it will be like tagging a name directly instead of a name in a sentence. (I guess its understood what i meant). That is the reason why im confused. 
Or other thought i had is instead of training a blank model, apply the pretrained model and the entities which are not recognized by that model can be trained from our data. Then combine those new entities with the entities in pre trained model and work together.

**Reply by polm** [reply]
> This is the data from one url. There is more than 5000 data from different urls which follow different pattern and length. I thought of considering data from one url as a single data and from other urls as different data. I need to extract similar entities ( not just these) like artist name, role etc etc and most of them dont follow a line format as we can see in the website.

That makes sense.

> Also how can i remove the lines which are irrelevant and filter those lines.

You'll need to look at your example websites and decide how to do that. Usually one or two word lines, or lines that contain certain phrase like "copyright" or "click for details" or something will be noise. 

> If i do that then it will be like tagging a name directly instead of a name in a sentence. (I guess its understood what i meant).

That's like if your input is just "John Smith" and you want to recognize it as a name, right? That's hard and it's not how NER is intended to be used. Technically your current documents have context, but if the show title or venue appears on a line by itself without any nearby explanation then it's going to be hard for the NER model to learn to recognize it. It's not impossible - if you have a line like "ASDF Auditorium" then the model can learn to use "Auditorium" as a hint - but it's more difficult than if you have complete sentences. For show titles there typically won't be keywords that can be used as hints like that so it's even more difficult.

> Or other thought i had is instead of training a blank model, apply the pretrained model and the entities which are not recognized by that model can be trained from our data. Then combine those new entities with the entities in pre trained model and work together.

That's a great way to get started. Take a look at the [double NER example project](https://github.com/explosion/projects/tree/v3/tutorials/ner_double).

-----

To be clear, you can try training a model on the data like you have it now, but since it looks like many of your entities have no useful context, I suspect the model will not work very well.

**Reply by aniyyanz08** [reply]
Thanks a lot for the valuable comments. I am going to proceed by considering all the observations from your side along with the data i have in hand. :)

**Reply by aniyyanz08** [reply]
>Technically your current documents have context, but if the show title or venue appears on a line by itself without any nearby explanation then it's going to be hard for the NER model to learn to recognize it. It's not impossible - if you have a line like "ASDF Auditorium" then the model can learn to use "Auditorium" as a hint - but it's more difficult than if you have complete sentences. For show titles there typically won't be keywords that can be used as hints like that so it's even more difficult.

If any of the entities appear in a line by itself, then should i create a sentence based on that entity. Or is there any other method to sort out such issues. There can be many data appearing like these.

**Reply by polm** [reply]
> If any of the entities appear in a line by itself, then should i create a sentence based on that entity. Or is there any other method to sort out such issues. There can be many data appearing like these.

To clarify one detail here, if there's something like this it's OK:

> Tonight's Show:
> A Very Whatever Christmas

The first line is context - it suggests the second line is a title - so the model can learn from it. The problem isn't strictly being on a line by itself, it's just that, in your example document, lines don't seem to refer to each other like this much, so you don't have any useful context.

If you don't have any context at all you can't create it out of thin air, so it may just be a hard problem. On the other hand, if your data is originally HTML, you may have other options if you get creative. There may be a way to integrate HTML tags or CSS classes which can be used as context. I'm not familiar with that process but I think it's been done before. A quick search turns up many systems and a good amount of research, like [this paper from last year](https://api.semanticscholar.org/CorpusID:227230378). 

If this is starting to sound like a lot of work, I'd encourage you to go ahead and train a CPU model on your data and see how it does, since you already have the annotations. Maybe hard cases like this are actually not that common in your dataset and you don't have to worry about this, but if it turns out they're hard at least you'll know what F1 score you're trying to beat.

**Reply by aniyyanz08** [reply]
After annotating the data and converting to spacy training data format, is it possible to visualize the annotation tags along with the training data using displacy. Just to check if the annotated results are correct after creating the spacy training set it will be helpful.

**Reply by polm** [reply]
Yes, that should just work. If you load the DocBin back in you get Docs, which can just be passed to displaCy.

**Reply by aniyyanz08** [reply]
>If this is starting to sound like a lot of work, I'd encourage you to go ahead and train a CPU model on your data and see how it does, since you already have the annotations. Maybe hard cases like this are actually not that common in your dataset and you don't have to worry about this, but if it turns out they're hard at least you'll know what F1 score you're trying to beat.

I have trained on CPU and getting a result as 

```
============================= Training pipeline =============================
ℹ Pipeline: ['tok2vec', 'ner']
ℹ Initial learn rate: 0.001
E    #       LOSS TOK2VEC  LOSS NER  ENTS_F  ENTS_P  ENTS_R  SCORE 
---  ------  ------------  --------  ------  ------  ------  ------
  0       0          0.00    162.59    0.00    0.00    0.00    0.00
 10     200       6939.94   9531.33   26.74   34.72   21.74    0.27
 20     400       1354.13   1546.09   90.67   92.73   88.70    0.91
 30     600       1075.98    731.88   94.55   99.05   90.43    0.95
 40     800        320.24    290.59   97.32  100.00   94.78    0.97
 50    1000         49.99    133.78   97.82   98.25   97.39    0.98
 60    1200        273.73    137.89   98.68  100.00   97.39    0.99
 70    1400         36.99     97.90   98.23  100.00   96.52    0.98
 80    1600         38.37     88.86   95.69   94.87   96.52    0.96
 90    1800        116.04    149.72   98.28   97.44   99.13    0.98
100    2000         52.27    103.25   99.14   98.29  100.00    0.99
110    2200         64.81     82.68   98.23  100.00   96.52    0.98
120    2400         81.99    121.73   95.69   94.87   96.52    0.96
130    2600        143.22    138.63   96.10   95.69   96.52    0.96
140    2800        173.41    144.73   98.28   97.44   99.13    0.98
150    3000         76.29     92.75   97.35   99.10   95.65    0.97
160    3200         23.69     50.71   98.68  100.00   97.39    0.99
170    3400         31.44     45.99   98.68  100.00   97.39    0.99
180    3600         53.77     55.34   99.14   98.29  100.00    0.99
```

Is it possible to give an insight on these values.

**Reply by polm** [reply]
You scores are basically perfect. If everything is configured correctly, that could be great news and mean you can use your model as-is. 

Honestly they seem a little too perfect though. Are you using the same data for train and dev? Do you have a lot of variation in your data? What's the output of `spacy debug data`?

**Reply by aniyyanz08** [reply]
Yes i have applied same data for training and dev. I didnt the whole data for training. I will check with more training data and separate one as dev. Thanks

**Reply by aniyyanz08** [reply]
@polm I have a sample data annotated. I will share the image for easy understanding

![tag](https://user-images.githubusercontent.com/92715688/140304971-af45e9cd-2330-4f16-9dac-73866bdcddbb.png)

here you can see the tagged data. The tagged data is the required ones in that page. (This data is obtained from a single url page as discussed in the thread above)
 My questions are: 
1. Rocco variations  are tagged as production_name in first line. In the data Rocco Variations is repeating in second line and can repeat somewhere in that document itself. So should we have to tag all the Rocco Variations in the document as production_name or one tag will be enough
2. **Steven Jarvi, Condutor** is not associated with that data. But Steven Jarvi is a **name** tag and conductor is another tag called creative team role. So should we have to tag that also eventhough it is not relevant to this document but will be relevant in another data.
3. Ann Arbor Orchestra is again repeating which comes to my doubt numbered 1

**Reply by polm** [reply]
The model will learn to reproduce the data it sees in training. That almost always means you should annotate every instance of something, so you should mark "Rococo Variations" in all locations. If you don't do that then your model will have conflicting information - sometimes "Rococo Variations" should be labelled, sometimes it shouldn't. How will the model tell the difference? Is there a reason only the first one is tagged in your data?

If you're lucky, the model will learn something like "production names occur on a line by themselves". If you're unlucky it will just behave strangely and unreliably.

Regarding the name/role, it is not clear to me why it is not relevant in this document. If it's not clear to a person, how will it be clear to the model? It's not impossible for the model to make those kinds of distinctions but it sounds hard.

**Reply by aniyyanz08** [reply]
There is no reason that i tagged this  "Rococo Variations" only once. I thought it will be fine by tagging once.
This url data is giving more details about Rocco Variations and details about that. As the data is a scraped data, you know the data can point to other pages or details in otherpages. So Steven Jarvi/Conductor is associated with another page. But as i understand from you reply, we have to tag that also. **So conclusion is : Whatever be the data, if anything comes related to our tag in that data, we have to tag those all, even multiple times is necessary if repeatedly comes.**

**Reply by aniyyanz08** [reply]
@polm According to the suggestion given, i have retagged the data. So can you have a quick glance at the tagged data and give a comment on the new tagging.

![newtag](https://user-images.githubusercontent.com/92715688/140865306-ebc48337-8aa2-4018-8199-35fcf33ea9e8.png)

One more doubt is if we tagged Steven Jarvi as a person, then some other parts the same name can be mentioned by just steven or jarvi. So if tagging **steven jarvi**, **steven a**nd **jarvi** as persons, will that cause any issue ?

**Reply by polm** [reply]
Those annotations look good.

> One more doubt is if we tagged Steven Jarvi as a person, then some other parts the same name can be mentioned by just steven or jarvi. So if tagging steven jarvi, steven and jarvi as persons, will that cause any issue ?

It's fine to tag all those as PERSON. 

The most important thing is that your input should be tagged the way you want the model to tag your output. Since "Steven Jarvi", "Steven", and "Jarvi" all refer to a person, they should all be tagged PERSON. This is not always true of substrings - for example, "Twelve Days of Christmas" could be a PRODUCTION, but "Christmas", "Days", and even "of" are of course not PRODUCTION on their own.

**Comment by shrinidhin** [comment]
Hi. The Dev data given while training the ner model for computing the accuracy of the model should also be tagged data right?

**Reply by polm** [reply]
Yes, the dev data needs labels. 

I'm not sure why you commented on this issue instead of opening a new one. If in doubt about whether your question is related to an existing discussion, it's often better to open a new thread than commenting on another one, especially if the older thread isn't currently active. (Keep in mind that when you comment on a thread, other participants in the thread usually get notifications.)

**Reply by shrinidhin** [reply]
Sure. I'll remember that. Thanks!
